File: ActionUIAndroid/library/src/main/java/com/abracode/actionui/Helpers/MapContract.kt
package com.abracode.actionui.Helpers

import com.abracode.actionui.Common.ActionUIElement
import com.abracode.actionui.Common.ActionUILogger
import com.abracode.actionui.Common.LoggerLevel
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull

/**
 * The provider-independent half of the `Map` element: Apple's JSON contract
 * (`ActionUI/Views/Map.swift`), validated once here so every Map provider
 * module renders the same documents identically.
 *
 * The element itself does NOT ship in the core library - a map engine is
 * distribution baggage for clients that never show one. Instead a client
 * links exactly one **provider module** (`:map-osm` - Leaflet/OpenStreetMap
 * in the platform WebView, dependency- and key-free; or `:map-google` -
 * the native Google map via maps-compose, Play Services + API key), which
 * self-registers its `ActionUIViewConstruction` as `"Map"` at app startup.
 * What the providers share - this contract parsing, plus the
 * [ActionUIValueType.COORDINATE] value bridge and [ActionUICoordinate] -
 * lives here in core, public, because a provider module cannot add a value
 * type or registry semantics from outside.
 *
 * [com.abracode.actionui.Common.ActionUIValueType.COORDINATE]
 */

/** One validated annotation: a pin with an optional title/subtitle popup. */
data class MapAnnotation(
    val coordinate: ActionUICoordinate,
    val title: String? = null,
    val subtitle: String? = null,
)

/** The `Map` element's validated properties, resolved once per `properties` change. */
data class MapConfig(
    val coordinate: ActionUICoordinate? = null,
    val showsUserLocation: Boolean = false,
    val interactionModes: List<String> = listOf("pan", "zoom", "rotate"),
    val annotations: List<MapAnnotation> = emptyList(),
    val actionID: String? = null,
    val valueChangeActionID: String? = null,
)

private val VALID_MAP_MODES = listOf("pan", "zoom", "rotate")

/**
 * The seed for the `Map` element's [com.abracode.actionui.Common.ViewModel]:
 * the `coordinate` property, or the zero coordinate. Providers return this
 * from `initialValue`, mirroring the Apple `Map.initialValue`.
 */
fun mapInitialCoordinate(element: ActionUIElement): ActionUICoordinate =
    element.properties?.coordinateProperty("coordinate") ?: ActionUICoordinate(0.0, 0.0)

/**
 * Resolves and validates the map properties, mirroring the Apple
 * `Map.validateProperties` warnings: an invalid `coordinate` is dropped, a
 * non-Boolean `showsUserLocation` defaults to false, malformed
 * `interactionModes` default to all, and annotations are validated
 * per-entry (an invalid coordinate skips the annotation; a non-String
 * title/subtitle is dropped). Pure (logging aside) so it is unit-testable.
 */
fun resolveMapConfig(props: JsonObject?, logger: ActionUILogger?): MapConfig {
    if (props == null) return MapConfig()

    var coordinate: ActionUICoordinate? = null
    if (props["coordinate"] != null) {
        coordinate = props.coordinateProperty("coordinate")
        if (coordinate == null) {
            logger?.log(
                "Map coordinate must be a dictionary with latitude/longitude numeric values",
                LoggerLevel.warning,
            )
        }
    }

    var showsUserLocation = false
    if (props["showsUserLocation"] != null) {
        val flag = props.booleanProperty("showsUserLocation")
        if (flag != null) {
            showsUserLocation = flag
        } else {
            logger?.log("Map showsUserLocation must be a Boolean; defaulting to false", LoggerLevel.warning)
        }
    }

    var modes = VALID_MAP_MODES
    val rawModes = props["interactionModes"]
    if (rawModes is JsonArray) {
        val names = rawModes.map { (it as? JsonPrimitive)?.takeIf { p -> p.isString }?.contentOrNull }
        if (names.all { it != null && it in VALID_MAP_MODES }) {
            modes = names.filterNotNull()
        } else {
            logger?.log(
                "Map interactionModes must be an array of 'pan', 'zoom', 'rotate'; defaulting to all",
                LoggerLevel.warning,
            )
        }
    } else if (rawModes != null) {
        logger?.log("Map interactionModes must be an array; defaulting to all", LoggerLevel.warning)
    }

    val annotations = mutableListOf<MapAnnotation>()
    val rawAnnotations = props["annotations"]
    if (rawAnnotations is JsonArray) {
        rawAnnotations.forEach { entry ->
            val dict = entry as? JsonObject
            val annotationCoordinate = dict?.coordinateProperty("coordinate")
            if (annotationCoordinate == null) {
                logger?.log("Map annotation coordinate invalid; skipping annotation", LoggerLevel.warning)
                return@forEach
            }
            var title: String? = null
            if (dict["title"] != null) {
                title = (dict["title"] as? JsonPrimitive)?.takeIf { it.isString }?.contentOrNull
                if (title == null) {
                    logger?.log("Map annotation title must be a String; defaulting to nil", LoggerLevel.warning)
                }
            }
            var subtitle: String? = null
            if (dict["subtitle"] != null) {
                subtitle = (dict["subtitle"] as? JsonPrimitive)?.takeIf { it.isString }?.contentOrNull
                if (subtitle == null) {
                    logger?.log("Map annotation subtitle must be a String; defaulting to nil", LoggerLevel.warning)
                }
            }
            annotations.add(MapAnnotation(annotationCoordinate, title, subtitle))
        }
    } else if (rawAnnotations != null) {
        logger?.log("Map annotations must be an array of dictionaries; defaulting to empty", LoggerLevel.warning)
    }

    return MapConfig(
        coordinate = coordinate,
        showsUserLocation = showsUserLocation,
        interactionModes = modes,
        annotations = annotations,
        actionID = props.stringProperty("actionID"),
        valueChangeActionID = props.stringProperty("valueChangeActionID"),
    )
}

File: ActionUIWeb/src/Helpers/MapContract.js
// MapContract.js — the provider-independent half of the `Map` element.
// Web port of ActionUIAndroid Helpers/MapContract.kt (+ CoordinateHelper.kt) and
// the shared contract Apple validates inline in ActionUI/Views/Map.swift. Canonical
// schema: Documentation/Schemas/Map.md.
//
// `Map` is provider-pluggable, the Android "link exactly one provider module"
// model: a provider self-registers as the "Map" element (register("Map", ...)) and
// owns the rendering, while THIS file owns the provider-independent contract that
// every provider must honor identically — the JSON validation and the coordinate
// value bridge. It registers nothing.
//
// Value bridge (valueType "string", Apple's contract): the value is the map's
// current center as a JSON coordinate string {"latitude":..,"longitude":..} (keys
// sorted, matching Apple's serializeValueToString .sortedKeys). A host
// setString re-centers; a user pan reports the new center back (a JS provider) and
// fires valueChangeActionID.
//
// Validation lives here and runs ONCE, at build time, inside each provider's
// buildView (resolveMapConfig) — the "validate where you read" discipline, so the
// warnings fire exactly once and a provider author calls one function. The web
// `validateProperties` for a Map provider is therefore a pass-through; baseline
// View modifiers (frame, cornerRadius, ...) still flow through it untouched.

// Parses a {latitude, longitude} dictionary into a coordinate, or null. Both
// fields must be finite numbers (Apple's double(forKey:) rejects strings). Mirrors
// Map.swift's coordinate(forKey:) and MapContract.kt's coordinateProperty.
export function parseCoordinate(dict) {
    if (dict === null || typeof dict !== "object" || Array.isArray(dict)) return null;
    const { latitude, longitude } = dict;
    if (typeof latitude !== "number" || !Number.isFinite(latitude)) return null;
    if (typeof longitude !== "number" || !Number.isFinite(longitude)) return null;
    return { latitude, longitude };
}

// Serializes a coordinate to the value-bridge string, keys sorted so `latitude`
// precedes `longitude` (Apple's .sortedKeys). The web's element value IS this
// string (valueType "string").
export function coordinateToValueString(coordinate) {
    return JSON.stringify({ latitude: coordinate.latitude, longitude: coordinate.longitude });
}

// Parses a value-bridge string back into a coordinate, or null on malformed input
// (bad JSON, missing/non-numeric fields, a non-object). Mirrors Map.swift's
// parseStringValue. Does not log — callers decide whether a bad host write warns.
export function parseCoordinateValueString(value) {
    if (typeof value !== "string") return null;
    let parsed;
    try {
        parsed = JSON.parse(value);
    } catch {
        return null;
    }
    return parseCoordinate(parsed);
}

// Coordinate equality (the value-bridge echo guard: tell a host write apart from
// the element's own report-back, the lastTracked pattern).
export function coordinatesEqual(a, b) {
    if (!a || !b) return a === b;
    return a.latitude === b.latitude && a.longitude === b.longitude;
}

// The seed for a Map's value: the `coordinate` property, or the zero coordinate.
// Mirrors mapInitialCoordinate (Kotlin) / Map.initialValue (Swift). Does not log.
export function mapInitialCoordinate(properties) {
    return parseCoordinate(properties?.coordinate) ?? { latitude: 0.0, longitude: 0.0 };
}

export const VALID_INTERACTION_MODES = ["pan", "zoom", "rotate"];

// Resolves and validates the Map properties into a typed config, mirroring
// Map.swift's validateProperties / MapContract.kt's resolveMapConfig warnings
// verbatim: an invalid coordinate is dropped, a non-Boolean showsUserLocation
// defaults to false, malformed interactionModes default to all, and annotations
// are validated per-entry (an invalid coordinate skips the annotation; a
// non-String title/subtitle is dropped). Pure aside from logging.
export function resolveMapConfig(properties, logger) {
    const config = {
        coordinate: null,
        showsUserLocation: false,
        interactionModes: VALID_INTERACTION_MODES.slice(),
        annotations: [],
        actionID: typeof properties?.actionID === "string" ? properties.actionID : null,
        valueChangeActionID: typeof properties?.valueChangeActionID === "string" ? properties.valueChangeActionID : null,
    };
    if (!properties || typeof properties !== "object") return config;

    if (properties.coordinate !== undefined) {
        const coordinate = parseCoordinate(properties.coordinate);
        if (coordinate) {
            config.coordinate = coordinate;
        } else {
            logger?.log("Map coordinate must be a dictionary with latitude/longitude numeric values", "warning");
        }
    }

    if (properties.showsUserLocation !== undefined) {
        if (typeof properties.showsUserLocation === "boolean") {
            config.showsUserLocation = properties.showsUserLocation;
        } else {
            logger?.log("Map showsUserLocation must be a Boolean; defaulting to false", "warning");
        }
    }

    if (properties.interactionModes !== undefined) {
        const modes = properties.interactionModes;
        if (Array.isArray(modes)) {
            if (modes.every((mode) => VALID_INTERACTION_MODES.includes(mode))) {
                config.interactionModes = modes.slice();
            } else {
                logger?.log("Map interactionModes must be an array of 'pan', 'zoom', 'rotate'; defaulting to all", "warning");
            }
        } else {
            logger?.log("Map interactionModes must be an array; defaulting to all", "warning");
        }
    }

    if (properties.annotations !== undefined) {
        if (Array.isArray(properties.annotations)) {
            for (const entry of properties.annotations) {
                const coordinate = parseCoordinate(entry && typeof entry === "object" ? entry.coordinate : undefined);
                if (!coordinate) {
                    logger?.log("Map annotation coordinate invalid; skipping annotation", "warning");
                    continue;
                }
                let title = null;
                if (entry.title !== undefined) {
                    if (typeof entry.title === "string") title = entry.title;
                    else logger?.log("Map annotation title must be a String; defaulting to nil", "warning");
                }
                let subtitle = null;
                if (entry.subtitle !== undefined) {
                    if (typeof entry.subtitle === "string") subtitle = entry.subtitle;
                    else logger?.log("Map annotation subtitle must be a String; defaulting to nil", "warning");
                }
                config.annotations.push({ coordinate, title, subtitle });
            }
        } else {
            logger?.log("Map annotations must be an array of dictionaries; defaulting to empty", "warning");
        }
    }

    return config;
}
